### PR TITLE
Trilinos/parallel reading

### PR DESCRIPTION
**Pull request by roigcarlo**
Parallel reading in memory can now be performed. 

Add the `"partition_system" : "Memory"` or `"partition_system" : "File"` to the  `"model_import_settings" ` in the  `"solver_settings" `

Default behavior if no option is provided is to do keep using files.

### CODE REVIEW

**Comment by roigcarlo** [comment]
It fixes #2789

File: applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py [hunk 1]
```diff
@@ -40,26 +40,46 @@ def ImportModelPart(self):
                 self.settings["model_import_settings"]["perform_partitioning"].SetBool(True)
 
             perform_partitioning = self.settings["model_import_settings"]["perform_partitioning"].GetBool()
+            perform_partitioning_in_memory = "File"
```
Comment (philbucher) [inline applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]: can you please use `file` and `memory`?

File: applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py [hunk 2]
```diff
@@ -40,26 +40,46 @@ def ImportModelPart(self):
...
+
+                    partitioner = KratosMetis.MetisDivideHeterogeneousInputInMemoryProcess(model_part_io, serial_model_part_io, number_of_partitions , domain_size, verbosity, sync_conditions)
                     partitioner.Execute()
+                    serial_model_part_io.ReadModelPart(self.main_model_part)
 
-                    KratosMultiphysics.Logger.PrintInfo("::[TrilinosImportModelPartUtility]::", "Metis divide finished.")
+                    if KratosMPI.mpi.rank == 0:
+                        KratosMultiphysics.Logger.PrintInfo("::[TrilinosImportModelPartUtility]::", "Metis divide finished.")
+                else:
+                    raise Exception("Invalid reading method selected")
```
Comment (philbucher) [inline applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]: `raise Exception('Invalid "partitioning_method" selected, possibilities are "file" and "memory"')`

File: applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py [hunk 3]
```diff
@@ -40,26 +40,46 @@ def ImportModelPart(self):
                 self.settings["model_import_settings"]["perform_partitioning"].SetBool(True)
 
             perform_partitioning = self.settings["model_import_settings"]["perform_partitioning"].GetBool()
+            perform_partitioning_in_memory = "File"
+
+            # Select the partitioning method
+            if self.settings["model_import_settings"].Has("partition_system"):
```
Comment (philbucher) [inline applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]: I think `partitioning_method` is a better name

**Review by philbucher — COMMENTED** [review]
some comments

File: applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py (hunk 1 above)
Comment (roigcarlo) [reply applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]: Sure, done

File: applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py (hunk 2 above)
Comment (roigcarlo) [reply applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]: :+1:

File: applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py [hunk 4]
```diff
@@ -40,26 +40,46 @@ def ImportModelPart(self):
                 self.settings["model_import_settings"]["perform_partitioning"].SetBool(True)
 
             perform_partitioning = self.settings["model_import_settings"]["perform_partitioning"].GetBool()
+            perform_partitioning_in_memory = "file"
```
Comment (jcotela) [inline applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]: Maybe `partition_type` or similar? The current name makes me want to answer you `true/false`.

File: applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h [hunk 5]
```diff
@@ -430,6 +431,7 @@ class MetisDivideHeterogeneousInputInMemoryProcess : public MetisDivideHeterogen
     ///@name Member Variables
     ///@{
 
+    IO& mrSerialIO;
```
Comment (jcotela) [inline applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h]: Shouldn't this be a `ModelPartIO&`? The only time it is used is through a downcasting.

Comment (roigcarlo) [reply applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h]: Actually, agree. We wanted to preserve the IO class for the "regular", but this one can be changed.

**Comment by roigcarlo** [comment]
About the names, I would like both to be the same 

I also though about `partitioning_method` name but that's not exactly what this change is about. We have several partition methods already (HeterogeneousDivideInput, GraphParitionig, Quadratic...), and eventually all of them could be based on memory.

But I agree that the name is not very descriptive...
Maybe something in the line of "Partitioning_physcial_support" o "partitioning_file". 

Also not disliking the idea of "true/false". Maybe it is simpler to change it to `"write_files": true/false`

File: applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py (hunk 3 above)
Comment (philbucher) [reply applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]: you forgot this one :)

Comment (philbucher) [reply applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]: sorry just saw your comment

**Comment by philbucher** [comment]
how abt 
- `partition_in_memory` => bool (if there are only two options)
- `partitioning_method` => HeterogeneousDivideInput, GraphParitionig, Quadratic ...
EDIT is memory/file partitioning available for all methods?
if not then you would throw the corresponding errors

File: applications/metis_application/custom_python/add_processes_to_python.cpp [hunk 6]
```diff
@@ -82,10 +82,10 @@ void AddProcessesToPython(pybind11::module& m)
...
     py::class_<MetisDivideHeterogeneousInputInMemoryProcess, MetisDivideHeterogeneousInputInMemoryProcess::Pointer, Process>(
         m,"MetisDivideHeterogeneousInputInMemoryProcess")
-        .def(py::init<IO&, unsigned int>())
-        .def(py::init<IO&, unsigned int, int>())
-        .def(py::init<IO&, unsigned int, int, int>())
-        .def(py::init<IO&, unsigned int, int, int, bool>())
+        .def(py::init<IO&, IO&, unsigned int>())
+        .def(py::init<IO&, IO&, unsigned int, int>())
+        .def(py::init<IO&, IO&, unsigned int, int, int>())
+        .def(py::init<IO&, IO&, unsigned int, int, int, bool>())
```
Comment (RiccardoRossi) [inline applications/metis_application/custom_python/add_processes_to_python.cpp]: Question: can't we generate a serial io internally?

**Comment by RiccardoRossi** [comment]
nice stuff btw ;-)

File: applications/metis_application/custom_python/add_processes_to_python.cpp (hunk 6 above)
Comment (roigcarlo) [reply applications/metis_application/custom_python/add_processes_to_python.cpp]: Mmmm... we could but then we have to pass the  modelpart to the process and the lecture of the memory buffer needs to be done inside the partitioning process.

File: applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py [hunk 7]
```diff
@@ -72,7 +90,8 @@ def ImportModelPart(self):
             self.settings["model_import_settings"]["input_filename"].SetString(mpi_input_filename)
 
             ## Read the new generated *.mdpa files
-            KratosMultiphysics.ModelPartIO(mpi_input_filename).ReadModelPart(self.main_model_part)
+            if perform_partitioning_in_memory == "file":
```
Comment (philbucher) [inline applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]:
```suggestion
            if not partition_in_memory:
```

File: applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py [hunk 8]
```diff
@@ -41,25 +41,43 @@ def ImportModelPart(self):
...
+                model_part_io = KratosMultiphysics.ReorderConsecutiveModelPartIO(input_filename)
+
+                if not partition_in_memory:
+                    ## Serial partition of the original .mdpa file
+                    if KratosMPI.mpi.rank == 0:
+                        partitioner = KratosMetis.MetisDivideHeterogeneousInputProcess(model_part_io, number_of_partitions , domain_size, verbosity, sync_conditions)
+                        partitioner.Execute()
+
+                        KratosMultiphysics.Logger.PrintInfo("::[TrilinosImportModelPartUtility]::", "Metis divide finished.")
+                else partition_in_memory:
```
Comment (philbucher) [inline applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]:
```suggestion
                else:
```

**Review by philbucher — COMMENTED** [review]
I think your current version does not run

**Comment by roigcarlo** [comment]
@philbucher your right!

File: applications/metis_application/custom_python/add_processes_to_python.cpp (hunk 6 above)
Comment (RiccardoRossi) [reply applications/metis_application/custom_python/add_processes_to_python.cpp]: ok, just forget about this

**Comment by RiccardoRossi** [comment]
we should use this in at least some of the tests...

**Comment by roigcarlo** [comment]
Ok I will prepare some

**Comment by roigcarlo** [comment]
I was trying to use some of the test from the trilinos app to make one for this feature but all the tests reading a modelpart are reporting errors or simply crashing. Is it normal?
I tried with these two:
- test_trilinos_redistance.py - Fails, also Crashes if I change the name to the one the process expects
- test_trilinos_levelset_convection.py - Crahses.

Run with openmpi 3.1.2 (mpirun -np 2 ...)

I can upload a simple test of my own in case they are known failures.
@jcotela @philbucher ?

**Comment by roigcarlo** [comment]
Also, the tests are not being installed. I will push a fix for that along with the tests...

**Comment by philbucher** [comment]
I think the tests broke with the Model_V3 branch, but I am not sure
Only remember that I also had problems running them
I did not write them though, so I cannot say much more ...

**Comment by roigcarlo** [comment]
The test problems should be fixed in #3343

**Review by RiccardoRossi — DISMISSED** [review]

**Comment by roigcarlo** [comment]
@RiccardoRossi I extended all the tests that were using the partitioner with the memory one. Redistance still gives me segmentation faults :(

I've also made some extra changes in the tests, please take a look at them.

**Review by RiccardoRossi — APPROVED** [review]

**Comment by philbucher** [comment]
just to add:
this is quite nice for making mpi-tests, this way there are no leftover files from partitioning

**Pull request merged**